File: r/R/result.R
# SPDX-FileCopyrightText: 2026 Arnaldo Mendes Pires Junior
# SPDX-License-Identifier: MIT

#' Build an optimization result object
#' @keywords internal
make_result <- function(x, fun, nit, nfev, success, message, direction) {
  structure(
    list(
      x = x,
      fun = fun,
      nit = nit,
      nfev = nfev,
      success = success,
      message = message,
      direction = direction,
      termination = infer_termination_reason(message)
    ),
    class = "givp_result"
  )
}

#' Infer termination reason from message
#' @keywords internal
infer_termination_reason <- function(message) {
  m <- tolower(message)
  if (grepl("converg", m)) return("converged")
  if (grepl("time", m)) return("time_limit_reached")
  if (grepl("early|stagn", m)) return("early_stop")
  if (grepl("feasible", m)) return("no_feasible")
  if (grepl("iter|max", m)) return("max_iterations_reached")
  "unknown"
}

#' Print optimization result
#' @param x A `givp_result` object.
#' @param ... Unused.
#' @export
print.givp_result <- function(x, ...) {
  cat("<givp_result>\n")
  cat("  success   :", x$success, "\n")
  cat("  fun       :", format(x$fun, digits = 8), "\n")
  cat("  nit       :", x$nit, "\n")
  cat("  nfev      :", x$nfev, "\n")
  cat("  direction :", x$direction, "\n")
  cat("  message   :", x$message, "\n")
  invisible(x)
}

File: rust/src/result.rs
// SPDX-FileCopyrightText: 2026 Arnaldo Mendes Pires Junior
// SPDX-License-Identifier: MIT

use crate::config::Direction;
use std::collections::HashMap;

/// Why the optimizer stopped.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub enum TerminationReason {
    Converged,
    MaxIterationsReached,
    TimeLimitReached,
    EarlyStop,
    NoFeasible,
    Unknown,
}

impl TerminationReason {
    pub fn from_message(msg: &str) -> Self {
        let lower = msg.to_lowercase();
        if lower.contains("converged") {
            Self::Converged
        } else if lower.contains("max") && lower.contains("iteration") {
            Self::MaxIterationsReached
        } else if lower.contains("time") {
            Self::TimeLimitReached
        } else if lower.contains("early") || lower.contains("stagnation") {
            Self::EarlyStop
        } else if lower.contains("no feasible") || lower.contains("no_feasible") {
            Self::NoFeasible
        } else {
            Self::Unknown
        }
    }
}

/// Optimization result.
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct OptimizeResult {
    /// Best solution found.
    pub x: Vec<f64>,
    /// Best objective value.
    pub fun: f64,
    /// Number of iterations.
    pub nit: usize,
    /// Number of function evaluations.
    pub nfev: usize,
    /// Whether the optimization succeeded.
    pub success: bool,
    /// Human-readable termination message.
    pub message: String,
    /// Optimization direction used.
    pub direction: Direction,
    /// Termination reason.
    pub termination: TerminationReason,
    /// Extra metadata.
    pub meta: HashMap<String, String>,
}

impl OptimizeResult {
    pub(crate) fn new(direction: Direction) -> Self {
        Self {
            x: Vec::new(),
            fun: f64::INFINITY,
            nit: 0,
            nfev: 0,
            success: true,
            message: String::new(),
            direction,
            termination: TerminationReason::Unknown,
            meta: HashMap::new(),
        }
    }
}
